### PR TITLE
GIT-810 & GIT-811: enforced checksum algorithms through configuration 

### LINKED ISSUE #810: Enforce the use of SHA256
**Problem to solve:**

Since [v1.3.1](https://github.com/blindsidenetworks/scalelite/releases/tag/v1.3.1), SL supports SHA256. The issue is that the implementation https://github.com/blindsidenetworks/scalelite/pull/697 allows the use of SHA1 and SHA256 together.

There are cases where the organization requires the strict use of only SHA256 (or even higher). In these cases SL should only accept the one defined.

**Purposed solution:**

Add a way to establish if SL will support multiple SHA algorithms (by default), and if it does not, only work with the one defined.

**Considered alternatives:**

**Additional context:**

### PR DESCRIPTION
**Pull request by jfederico**
Closes #810

### CODE REVIEW

File: app/controllers/concerns/api_helper.rb [hunk 1]
```diff
@@ -11,29 +11,38 @@ module ApiHelper
...
     check_string += request.query_string.gsub(
       /&checksum=#{params[:checksum]}|checksum=#{params[:checksum]}&|checksum=#{params[:checksum]}/, ''
     )
 
-    return if Rails.configuration.x.loadbalancer_secrets.any? do |secret|
-      checksum_sha1 = Digest::SHA1.hexdigest(check_string + secret)
-      checksum_sha256 = Digest::SHA256.hexdigest(check_string + secret)
-      ActiveSupport::SecurityUtils.secure_compare(checksum_sha1, params[:checksum]) ||
-      ActiveSupport::SecurityUtils.secure_compare(checksum_sha256, params[:checksum])
+    checksum_algorithms = Rails.configuration.x.loadbalancer_checksum_algorithms
```
Comment (kepstin) [inline app/controllers/concerns/api_helper.rb:31]: Rather than trying all supported checksum algorithms, this should try only the selected one (based on the length of the provided checksum). It still needs a check that the provided checksum matches an enabled algorithm as well, of course.

File: app/controllers/concerns/api_helper.rb [hunk 2]
```diff
@@ -11,29 +11,38 @@ module ApiHelper
...
+    secrets.product(checksum_algorithms).any? do |secret, checksum_algorithm|
+      return true if ActiveSupport::SecurityUtils.secure_compare(get_checksum(check_string + secret, checksum_algorithm),
+      params[:checksum])
     end
 
     raise ChecksumError
   end
 
+  def get_checksum(check_string, checksum_algorithm)
+    return Digest::SHA512.hexdigest(check_string) if checksum_algorithm == 'SHA512'
```
Comment (kepstin) [inline app/controllers/concerns/api_helper.rb:41]: This should be reformatted as a `case` statement. Consider setting a default value for the `checksum_algorithm` option.

The method should raise an exception if an unsupported value of `checksum_algorithm` is passed.

File: app/controllers/concerns/api_helper.rb [hunk 3]
```diff
@@ -82,7 +91,7 @@ def valid_token?(token)
   def post_req(uri, body)
     http = Net::HTTP.new(uri.host, uri.port)
     http.use_ssl = (uri.scheme == 'https')
-    exp = Time.now.to_i + 24 * 3600
+    exp = Time.now.to_i + (24 * 3600)
```
Comment (kepstin) [inline app/controllers/concerns/api_helper.rb:94]: Unrelated change?

File: config/application.rb [hunk 4]
```diff
@@ -44,6 +44,10 @@ class Application < Rails::Application
     config.x.loadbalancer_secrets.push(ENV['LOADBALANCER_SECRET']) if ENV['LOADBALANCER_SECRET']
     config.x.loadbalancer_secrets.concat(ENV['LOADBALANCER_SECRETS'].split(':')) if ENV['LOADBALANCER_SECRETS']
 
+    # Algorithms used for calculating checksum [SHA1|SHA256|...]
+    config.x.loadbalancer_checksum_algorithm = ENV.fetch('LOADBALANCER_CHECKSUM_ALGORITHM', 'SHA1:SHA256:SHA512')
```
Comment (kepstin) [inline config/application.rb:48]: Needs a documentation update in the README.md file.

**Review by kepstin — CHANGES_REQUESTED** [review]
This work also needs some automated tests that ensure checksums from all enabled algorithms are accepted, and that checksums from disabled algorithms are rejected.

File: app/controllers/concerns/api_helper.rb (hunk 1 above)
Comment (jfederico) [reply app/controllers/concerns/api_helper.rb:31]: Agree. Although this review should have come before. This is how the application has been dealing with multiple algorithms for more than 1 year. We may revisit this once BBB does what it is expected

File: app/controllers/concerns/api_helper.rb (hunk 2 above)
Comment (jfederico) [reply app/controllers/concerns/api_helper.rb:41]: yeah, it would work that way too

File: app/controllers/concerns/api_helper.rb (hunk 3 above)
Comment (jfederico) [reply app/controllers/concerns/api_helper.rb:94]: Got caught by rubocop. If it is not changed it won't pass

File: config/application.rb (hunk 4 above)
Comment (jfederico) [reply config/application.rb:48]: Correct

**Comment by jfederico** [comment]
Discarded in favour of the already merged PR #828
